Mathlib/Analysis/Fourier/ZMod.lean

```diff
@@ -0,0 +1,131 @@
+* `DirichletCharacter.fourierTransform_eq_inv_mul_gaussSum`: the discrete Fourier transform of a
+  primitive Dirichlet character `χ` is a Gauss sum times `χ⁻¹`.
+-/
+
+open scoped Real
+
+namespace ZMod
+
+variable {N : ℕ} [NeZero N]
+
+/-- The `AddMonoidHom` from `ZMod N` to `ℝ / ℤ` sending `j mod N` to `j / N mod 1`. -/
+noncomputable def toAddCircle : ZMod N →+ UnitAddCircle :=
+  lift N ⟨AddMonoidHom.mk' (fun j ↦ ↑(j / N : ℝ)) (by simp [add_div]), by simp⟩
+
+lemma toAddCircle_coe (j : ℤ) :
+    toAddCircle (j : ZMod N) = ↑(j / N : ℝ) := by
+  simp [toAddCircle]
+
+lemma toAddCircle_apply (j : ZMod N) :
+    toAddCircle j = ↑(j.val / N : ℝ) := by
+  conv_lhs => rw [show j = (val j : ℤ) by simp, toAddCircle_coe]
+  simp only [natCast_val, intCast_cast]
+
+variable (N) in
+lemma toAddCircle_injective : Function.Injective (toAddCircle : ZMod N → _) := by
+  intro x y hxy
+  have : (0 : ℝ) < N := Nat.cast_pos.mpr (NeZero.pos _)
+  rwa [toAddCircle_apply, toAddCircle_apply, AddCircle.coe_eq_coe_iff_of_mem_Ico
+    (hp := Real.fact_zero_lt_one) (a := 0), div_left_inj' this.ne', Nat.cast_inj,
+    (val_injective N).eq_iff] at hxy <;>
+  exact ⟨by positivity, by simpa only [zero_add, div_lt_one this, Nat.cast_lt] using val_lt _⟩
+
+/-- The additive character from `ZMod N` to the unit circle in `ℂ`, sending `j mod N` to
+`exp (2 * π * I * j / N)`. -/
+noncomputable def toCircle : AddChar (ZMod N) circle where
+  toFun := fun j ↦ (toAddCircle j).toCircle
+  map_add_eq_mul' a b := by simp_rw [map_add, AddCircle.toCircle_add]
+  map_zero_eq_one' := by simp_rw [map_zero, AddCircle.toCircle, ← QuotientAddGroup.mk_zero,
+    Function.Periodic.lift_coe, mul_zero, expMapCircle_zero]
+
+lemma toCircle_coe (j : ℤ) :
+    toCircle (j : ZMod N) = Complex.exp (2 * π * Complex.I * j / N) := by
+  rw [toCircle, AddChar.coe_mk, AddCircle.toCircle, toAddCircle_coe,
+    Function.Periodic.lift_coe, expMapCircle_apply]
+  push_cast
+  ring_nf
+
+lemma toCircle_apply (j : ZMod N) :
+    toCircle j = Complex.exp (2 * π * Complex.I * j.val / N) := by
+  rw [← Int.cast_natCast, ← toCircle_coe, natCast_val, intCast_zmod_cast]
+
+/-- The additive character from `ZMod N` to `ℂ`, sending
+`j mod N` to `exp (2 * π * I * j / N)`. -/
+noncomputable def stdAddChar : AddChar (ZMod N) ℂ := circle.subtype.compAddChar toCircle
+
+lemma stdAddChar_coe (j : ℤ) :
+    stdAddChar (j : ZMod N) = Complex.exp (2 * π * Complex.I * j / N) := by
+  simp only [stdAddChar, MonoidHom.coe_compAddChar, Function.comp_apply,
+    Submonoid.coe_subtype, toCircle_coe]
+
+lemma stdAddChar_apply (j : ZMod N) : stdAddChar j = ↑(toCircle j) := rfl
+
+section fourier
+
+open MeasureTheory
+
+/-- The discrete measurable space structure (every set is measurable). -/
+instance : MeasurableSpace (ZMod N) := ⊤
+
+/-- The discrete topology (every set is open). -/
+instance : TopologicalSpace (ZMod N) := ⊥
+
+instance : DiscreteTopology (ZMod N) := ⟨rfl⟩
```

Comment:
Can all of these move to earlier files?

Reply:
[Fin.instMeasurableSpace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/MeasurableSpace/Basic.html#Fin.instMeasurableSpace) is in `MeasureTheory.MeasurableSpace.Basic`, so I guess the `MeasurableSpace (ZMod n)` instance could go there?

Reply:
That file's already 2200 lines long; we should be moving stuff out of it, not adding more.

Reply:
See [#13937](https://github.com/leanprover-community/mathlib4/pull/13937)

Reply:
Okay, what about now?

Reply:
Give me a few hours, I do have a day job too 😄